File: sdks/python/examples/basic_usage.py
"""
Basic usage example for AgentID Python SDK.

This example shows how to use the SDK to make authenticated
requests to services that accept AgentID credentials.
"""

import asyncio

from agentid import AgentCredential, AgentClient


async def main():
    # Example 1: Basic credential usage
    print("=== Example 1: Basic AgentCredential ===")

    cred = AgentCredential("cred_example")

    # Load credential (happens automatically on first request)
    try:
        await cred.load_async()
        print(f"Loaded credential for: {cred._credential_data.agent_name}")
        print(f"Expires in: {cred.time_to_expiry:.0f} seconds")
    except Exception as e:
        print(f"Error loading credential: {e}")
        # Continue with examples anyway for demonstration

    # Get headers to add to your own requests
    headers = cred.get_headers(
        method="GET",
        url="https://api.example.com/data",
    )
    print(f"Headers to include: {list(headers.keys())}")

    # Example 2: Making requests with AgentCredential
    print("\n=== Example 2: Making Requests ===")

    # Async request (uses httpx under the hood)
    # response = await cred.request_async("https://api.example.com/data")
    # print(f"Response status: {response.status_code}")

    # Sync request
    # response = cred.request("https://api.example.com/data", method="POST", json={"key": "value"})
    # print(f"Response: {response.json()}")

    print("(Skipped actual requests - uncomment to test with real API)")

    # Example 3: Using AgentClient
    print("\n=== Example 3: Using AgentClient ===")

    # AgentClient provides a higher-level interface
    # async with AgentClient("cred_example", base_url="https://api.example.com") as client:
    #     response = await client.get("/data")
    #     response = await client.post("/users", json={"name": "Test"})

    print("(Skipped - uncomment to test with real API)")

    # Example 4: Wrapping existing httpx client
    print("\n=== Example 4: Wrapping httpx Client ===")

    import httpx

    async with httpx.AsyncClient() as client:
        # Wrap the client to auto-inject headers
        cred.wrap_httpx(client)

        # Now all requests include AgentID headers
        # response = await client.get("https://api.example.com/data")

    print("(Skipped - uncomment to test with real API)")


if __name__ == "__main__":
    asyncio.run(main())

File: sdks/typescript/examples/basic-usage.ts
/**
 * Basic usage example for AgentID TypeScript SDK
 *
 * This example shows how to use the SDK to make authenticated
 * requests to services that accept AgentID credentials.
 */

import { AgentCredential, AgentIDError } from '../src';

async function main() {
  // Example 1: Basic credential usage
  console.log('=== Example 1: Basic AgentCredential ===\n');

  const cred = new AgentCredential('cred_example');

  // Load credential (happens automatically on first request)
  try {
    await cred.load();
    console.log(`Loaded credential for: ${cred.data?.agent_name}`);
    console.log(`Expires in: ${Math.floor(cred.timeToExpiry / 1000)} seconds`);
  } catch (error) {
    if (error instanceof AgentIDError) {
      console.log(`Error loading credential: ${error.message}`);
    }
  }

  // Get headers to add to your own requests
  const headers = await cred.getHeaders('GET', 'https://api.example.com/data');
  console.log('Headers to include:', Object.keys(headers));

  // Example 2: Making requests with AgentCredential
  console.log('\n=== Example 2: Making Requests ===\n');

  // Using the built-in fetch wrapper
  // const response = await cred.fetch('https://api.example.com/data');
  // console.log(`Response status: ${response.status}`);

  // With POST and body
  // const response = await cred.fetch('https://api.example.com/users', {
  //   method: 'POST',
  //   headers: { 'Content-Type': 'application/json' },
  //   body: JSON.stringify({ name: 'New User' }),
  // });

  console.log('(Skipped actual requests - uncomment to test with real API)');

  // Example 3: Creating an authenticated fetch function
  console.log('\n=== Example 3: Authenticated Fetch ===\n');

  const authenticatedFetch = cred.createFetch();

  // Use it like regular fetch
  // const response = await authenticatedFetch('https://api.example.com/data');

  console.log('Created authenticated fetch function');

  // Example 4: Credential lifecycle
  console.log('\n=== Example 4: Credential Lifecycle ===\n');

  console.log(`Is loaded: ${cred.isLoaded}`);
  console.log(`Is active: ${cred.isActive}`);
  console.log(`Is expired: ${cred.isExpired}`);
  console.log(`Time to expiry: ${cred.timeToExpiry}ms`);
  console.log(`Needs refresh: ${cred.needsRefresh}`);
}

main().catch(console.error);
